File: LeetCode/Medium/0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int length = findLength(head);
        int i = 0, traverseTill = length - n - 1;
        if (traverseTill == -1) return head.next;
        ListNode cur = head;
        while (i < traverseTill) {
            cur = cur.next;
            i++;
        }
        cur.next = cur.next.next;
        return head;
    }

    public int findLength(ListNode head) {
        int count = 0;
        if (head == null) return count;
        ListNode cur = head;
        while (cur != null) {
            count++;
            cur = cur.next;
        }
        return count;
    }
}

File: LeetCode/Medium/0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        var size = 0
        var node = head
        while (node != null) {
            size++
            node = node.next
        }
        if (size == 1) {
            return null
        }
        if (size == 2) {
            if (n == 1) {
                head?.next = null
                return head
            }
            if (n == 2) {
                return head?.next
            }
            return head
        }
        if (size == n) {
            return head?.next
        }
        node = head
        for (i in 0..size - n - 2) {
            node = node?.next
        }
        node?.next = node?.next?.next
        return head
    }
}
